File: sdks/java/src/dev/ironauth/verify/SelfTest.java
// SPDX-License-Identifier: MIT OR Apache-2.0
package dev.ironauth.verify;

import dev.ironauth.verify.IronAuthVerifier.Reason;
import dev.ironauth.verify.IronAuthVerifier.VerifyException;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.interfaces.EdECPublicKey;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * The properties the shared corpus CANNOT express (issue #118).
 *
 * <p>The corpus is a fixed list of tokens signed by fixed keys, so every property it can state is
 * of the form "this exact token gets this answer". Three things this verifier claims are not of
 * that form, and each was chosen because a mutation run showed the corpus could not see it:
 *
 * <ul>
 *   <li><strong>Key injection is refused BEFORE the signature is checked.</strong> The corpus's
 *       {@code embedded_jwk_key_injection} vector carries an attacker's key and an attacker's
 *       signature, so a verifier with no structural check still refuses it -- the signature fails
 *       against the real key. Deleting the check keeps the corpus green. The claim that the
 *       refusal is <em>structural</em> therefore needs a token whose signature is genuinely VALID,
 *       which the corpus has no key to mint.
 *   <li><strong>Size is bounded before decoding.</strong> No fixed corpus contains a 9 KiB token.
 *   <li><strong>A token with no expiry never verifies.</strong> Every corpus vector has {@code exp}.
 * </ul>
 *
 * <p>Run as {@code java dev.ironauth.verify.SelfTest}; it exits non-zero and names each failure.
 */
public final class SelfTest {

    private SelfTest() {}

    private static final String ISSUER = "https://issuer.example/t/tnt_self/e/env_self";
    private static final String AUDIENCE = "cli_self";
    private static final long NOW = 1_800_000_000L;

    private static final List<String> FAILURES = new ArrayList<>();

    /**
     * How many properties actually ran.
     *
     * <p>Counted rather than written down. A hand-written total in the success line is a claim
     * about this file made somewhere else in it, and it goes stale the first time someone
     * comments an assertion out -- leaving a green run reporting a number that was true once.
     */
    private static int checked;

    public static void main(String[] args) throws Exception {
        KeyPair pair = KeyPairGenerator.getInstance("Ed25519").generateKeyPair();
        TrustedKey trusted = new TrustedKey("self-1", "OKP", pair.getPublic());
        IronAuthVerifier verifier =
                new IronAuthVerifier(Set.of("EdDSA"), List.of(trusted), ISSUER, AUDIENCE, 0);

        String claims = "{\"iss\":\"" + ISSUER + "\",\"aud\":\"" + AUDIENCE
                + "\",\"sub\":\"usr_self\",\"exp\":" + (NOW + 3600) + "}";

        // THE CONTROL, and it is the whole reason the next assertion means anything. It varies
        // from the injection token in exactly ONE respect: the header. Without it, a refusal
        // below could be caused by a mint this test got wrong, and would look identical.
        String honest = Fixtures.mint(pair, "{\"alg\":\"EdDSA\",\"typ\":\"JWT\",\"kid\":\"self-1\"}", claims);
        accepts(verifier, honest, "a correctly minted token verifies (the control for the next case)");

        // The property the corpus cannot reach: a VALID signature by a PUBLISHED key, with a
        // `jwk` header bolted on. A verifier that merely ignores the embedded key accepts this,
        // because the signature really does verify. Only a structural refusal catches it.
        String injected = Fixtures.mint(
                pair,
                "{\"alg\":\"EdDSA\",\"typ\":\"JWT\",\"kid\":\"self-1\",\"jwk\":{\"kty\":\"OKP\",\"crv\":\"Ed25519\",\"x\":\"F83SEmSVgKMBLYCoZfCPDHVGDGVoXVfyxRZsGnPPYQE\"}}",
                claims);
        refuses(verifier, injected, Reason.EMBEDDED_KEY_INJECTION,
                "a header carrying its own key is refused even when the signature is VALID");

        // The same, for the indirect forms. `jku` and `x5u` point at a key instead of carrying
        // one, which is the same attack with a fetch in the middle.
        for (String member : List.of("jku\":\"https://attacker.example/keys", "x5u\":\"https://attacker.example/chain")) {
            String header = "{\"alg\":\"EdDSA\",\"typ\":\"JWT\",\"kid\":\"self-1\",\"" + member + "\"}";
            refuses(verifier, Fixtures.mint(pair, header, claims), Reason.EMBEDDED_KEY_INJECTION,
                    "a header pointing AT a key is refused too, not only one carrying it");
        }

        // Bounded before decoding. A validly signed token padded past the cap must still be
        // refused on size, so the bound is not merely documented.
        String oversize = honest + "A".repeat(IronAuthVerifier.MAX_TOKEN_BYTES);
        refuses(verifier, oversize, Reason.MALFORMED_STRUCTURE, "an oversized token is refused before decoding");

        // Padding is not accepted: one token must have exactly one encoding.
        refuses(verifier, honest.substring(0, honest.lastIndexOf('.')) + ".AA==",
                Reason.BASE64_MALFORMED, "a PADDED segment is refused, so a token has one encoding");

        // A token with no expiry is a token that never expires.
        String noExpiry = Fixtures.mint(
                pair,
                "{\"alg\":\"EdDSA\",\"typ\":\"JWT\",\"kid\":\"self-1\"}",
                "{\"iss\":\"" + ISSUER + "\",\"aud\":\"" + AUDIENCE + "\",\"sub\":\"usr_self\"}");
        refuses(verifier, noExpiry, Reason.CLAIMS_MALFORMED, "a token with no exp never verifies");

        // NO INPUT ESCAPES AS AN UNCHECKED EXCEPTION.
        //
        // This is the property a review found broken. `verify` declares VerifyException, so a
        // caller writes one catch and believes it covers every bad token. Two inputs made the
        // hand-written JSON reader throw StringIndexOutOfBoundsException instead -- not an
        // IllegalArgumentException, so it sailed past the verifier's catch and out of the method.
        // In a servlet that is a 500 where a 401 belonged: the wrong status, the wrong alert, and
        // a stack trace in the log for what is simply an invalid token.
        //
        // The corpus cannot reach this. Its malformed vectors are malformed BASE64; these are
        // valid base64 carrying hostile JSON, and each one is a shape the reader has to bound.
        char backslash = (char) 92;
        String head = "{\"alg\":\"EdDSA\",\"kid\":\"self-1\",\"x\":\"";
        List<String> hostileHeaders = List.of(
                head + backslash,                          // input ends inside an escape
                head + backslash + "u12",                  // input ends inside a hex escape
                head + backslash + "uZZZZ" + "\"}",         // a hex escape with non-hex digits
                head + backslash + "q\"}",                  // an escape that does not exist
                "[".repeat(6100),                          // nesting deep enough to blow the stack
                "{\"a\":".repeat(6100) + "1");             // the same, through objects
        for (String hostile : hostileHeaders) {
            String token = Fixtures.b64(hostile.getBytes(java.nio.charset.StandardCharsets.UTF_8)) + ".e30.AA";
            checked++;
            try {
                verifier.verify(token, NOW);
                FAILURES.add("a hostile header VERIFIED: " + summarise(hostile));
            } catch (VerifyException expected) {
                // A refusal is the whole point; which refusal is not, since these are all garbage.
            } catch (Throwable escaped) {
                FAILURES.add("a hostile header threw " + escaped.getClass().getName()
                        + " out of verify(), which callers do not catch: " + summarise(hostile));
            }
        }

        // AND THE SAME PROPERTY ONE LAYER DOWN, on the parser itself.
        //
        // The loop above passes with EITHER the reader's bounds checks OR the verifier's
        // RuntimeException backstop in place, because each masks the other: a mutation run
        // removing just one left the suite green. Two layers is the right design here, but it
        // means neither is measured by an end-to-end test alone.
        //
        // So this asserts the reader's own contract directly, which only the bounds checks can
        // satisfy: `Json.parse` throws IllegalArgumentException and NOTHING else. The backstop
        // cannot make this pass, so the two layers now have one test each.
        for (String hostile : hostileHeaders) {
            checked++;
            try {
                Json.parse(hostile);
                FAILURES.add("Json.parse accepted hostile input: " + summarise(hostile));
            } catch (IllegalArgumentException expected) {
                // The declared failure mode.
            } catch (Throwable escaped) {
                FAILURES.add("Json.parse threw " + escaped.getClass().getName()
                        + ", which is outside its contract: " + summarise(hostile));
            }
        }

        // `nbf` present but not a number must be MALFORMED, not treated as absent: otherwise
        // `"nbf": "tomorrow"` silently disables the check it was written to perform.
        refuses(verifier, Fixtures.mint(pair, "{\"alg\":\"EdDSA\",\"typ\":\"JWT\",\"kid\":\"self-1\"}",
                        "{\"iss\":\"" + ISSUER + "\",\"aud\":\"" + AUDIENCE + "\",\"exp\":" + (NOW + 3600)
                                + ",\"nbf\":\"tomorrow\"}"),
                Reason.CLAIMS_MALFORMED, "a non-numeric nbf is malformed, not absent");

        // An empty allow-list reads as "allow nothing" and behaves as a silent outage, so it is
        // refused at construction rather than at the first request.
        try {
            checked++;
            new IronAuthVerifier(Set.of(), List.of(trusted), ISSUER, AUDIENCE, 0);
            FAILURES.add("an empty algorithm allow-list was accepted at construction");
        } catch (IllegalArgumentException expected) {
            // as intended
        }

        // The Ed25519 JWK encoding round-trips. The corpus proves DECODING works; this proves the
        // parity bit is handled in both directions, which is the half a fixed corpus cannot see.
        String x = Fixtures.encodeEd25519((EdECPublicKey) pair.getPublic());
        String jwks = "{\"keys\":[{\"kty\":\"OKP\",\"crv\":\"Ed25519\",\"x\":\"" + x + "\",\"kid\":\"self-1\"}]}";
        List<TrustedKey> decoded = TrustedKey.fromJwks(jwks);
        checked++;
        if (decoded.size() != 1 || !decoded.get(0).key().equals(pair.getPublic())) {
            FAILURES.add("an Ed25519 key did not survive a JWK round trip: " + decoded);
        }

        if (!FAILURES.isEmpty()) {
            System.err.println("FAIL: the Java verifier does not hold its own claims");
            FAILURES.forEach(failure -> System.err.println("  - " + failure));
            System.exit(1);
        }
        // A floor, so commenting assertions out fails here instead of reporting a smaller number
        // in a green run. It is a floor and not an equality: adding a property should not break it.
        if (checked < 22) {
            System.err.println("FAIL: only " + checked + " properties ran; this suite is its list");
            System.exit(1);
        }
        System.out.println("java self-test: " + checked + " properties the corpus cannot express OK");
    }

    /** A short, log-safe rendering of a hostile input, so a failure names it without a wall of text. */
    private static String summarise(String hostile) {
        String head = hostile.length() > 40 ? hostile.substring(0, 40) + "..." : hostile;
        return "(" + hostile.length() + " chars) " + head;
    }

    private static void accepts(IronAuthVerifier verifier, String token, String why) {
        checked++;
        try {
            verifier.verify(token, NOW);
        } catch (VerifyException refused) {
            FAILURES.add(why + " -- refused as " + refused.reason());
        }
    }

    private static void refuses(IronAuthVerifier verifier, String token, Reason expected, String why) {
        checked++;
        try {
            verifier.verify(token, NOW);
            FAILURES.add(why + " -- but it VERIFIED");
        } catch (VerifyException refused) {
            if (refused.reason() != expected) {
                FAILURES.add(why + " -- expected " + expected + ", got " + refused.reason());
            }
        }
    }

}

File: sdks/dotnet/IronAuth.Verify.Checks/SelfTest.cs
// SPDX-License-Identifier: MIT OR Apache-2.0
using System.Globalization;
using System.Text;
using IronAuth.Verify;
using Org.BouncyCastle.Crypto;

namespace IronAuth.Verify.Checks;

/// <summary>
/// The properties the shared corpus CANNOT express (issue #118).
/// </summary>
/// <remarks>
/// <para>
/// The corpus is a fixed list of tokens signed by fixed keys, so every property it can state has
/// the form "this exact token gets this answer". Several things this verifier claims are not of
/// that form, and each was chosen because a mutation run showed the corpus blind to it:
/// </para>
/// <para>
/// Key injection is refused BEFORE the signature is checked. The corpus's
/// <c>embedded_jwk_key_injection</c> vector carries an attacker's key AND an attacker's signature,
/// so a verifier with no structural check still refuses it: the signature fails against the real
/// key. Deleting the check keeps the corpus green. Proving the refusal is structural needs a token
/// whose signature is genuinely VALID, which the corpus has no private key to mint.
/// </para>
/// <para>
/// Size is bounded before decoding; no fixed corpus contains a 9 KiB token. A token with no
/// expiry never verifies; every corpus vector has <c>exp</c>. And no input escapes
/// <c>Verify</c> as an unexpected exception: the corpus's malformed vectors are malformed BASE64,
/// where these are valid base64 carrying hostile JSON, which is a different door.
/// </para>
/// </remarks>
internal static class SelfTest
{
    private const string Issuer = "https://issuer.example/t/tnt_self/e/env_self";
    private const string Audience = "cli_self";
    private const long Now = 1_800_000_000L;

    private static readonly List<string> Failures = [];
    private static int _checked;

    internal static int Run()
    {
        AsymmetricCipherKeyPair pair = Fixtures.GenerateEd25519();
        IReadOnlyList<TrustedKey> keys = TrustedKey.FromJwks(
            $$"""{"keys":[{"kty":"OKP","crv":"Ed25519","x":"{{Fixtures.PublicX(pair)}}","kid":"self-1"}]}""");
        IronAuthVerifier verifier = new(["EdDSA"], keys, Issuer, Audience, 0);

        string claims = string.Create(CultureInfo.InvariantCulture,
            $$"""{"iss":"{{Issuer}}","aud":"{{Audience}}","sub":"usr_self","exp":{{Now + 3600}}}""");

        // THE CONTROL, and the reason the next assertion means anything. It varies from the
        // injection token in exactly ONE respect: the header. Without it, a refusal below could be
        // caused by a mint this test got wrong, and would look identical.
        string honest = Fixtures.Mint(pair, """{"alg":"EdDSA","typ":"JWT","kid":"self-1"}""", claims);
        Accepts(verifier, honest, "a correctly minted token verifies (the control for the next case)");

        // The property the corpus cannot reach: a VALID signature by a PUBLISHED key with a `jwk`
        // header bolted on. A verifier that merely ignores the embedded key accepts this, because
        // the signature really does verify. Only a structural refusal catches it.
        string injected = Fixtures.Mint(
            pair,
            """{"alg":"EdDSA","typ":"JWT","kid":"self-1","jwk":{"kty":"OKP","crv":"Ed25519","x":"F83SEmSVgKMBLYCoZfCPDHVGDGVoXVfyxRZsGnPPYQE"}}""",
            claims);
        Refuses(verifier, injected, RejectReason.EmbeddedKeyInjection,
            "a header carrying its own key is refused even when the signature is VALID");

        // The indirect forms: `jku` and `x5u` point AT a key instead of carrying one, which is the
        // same attack with a fetch in the middle.
        foreach (string member in new[] { """jku":"https://attacker.example/keys""", """x5u":"https://attacker.example/chain""" })
        {
            string header = $$"""{"alg":"EdDSA","typ":"JWT","kid":"self-1","{{member}}"}""";
            Refuses(verifier, Fixtures.Mint(pair, header, claims), RejectReason.EmbeddedKeyInjection,
                "a header pointing AT a key is refused too, not only one carrying it");
        }

        // Bounded before decoding. A validly signed token padded past the cap must still be refused
        // on size, so the bound is not merely documented.
        Refuses(verifier, honest + new string('A', IronAuthVerifier.MaxTokenBytes),
            RejectReason.MalformedStructure, "an oversized token is refused before decoding");

        // Padding is not accepted: one token must have exactly one encoding.
        Refuses(verifier, honest[..honest.LastIndexOf('.')] + ".AA==",
            RejectReason.Base64Malformed, "a PADDED segment is refused, so a token has one encoding");

        // A token with no expiry is a token that never expires.
        Refuses(verifier,
            Fixtures.Mint(pair, """{"alg":"EdDSA","typ":"JWT","kid":"self-1"}""",
                $$"""{"iss":"{{Issuer}}","aud":"{{Audience}}","sub":"usr_self"}"""),
            RejectReason.ClaimsMalformed, "a token with no exp never verifies");

        // `nbf` present but not a number must be MALFORMED, not treated as absent: otherwise
        // `"nbf": "tomorrow"` silently disables the check it was written to perform.
        Refuses(verifier,
            Fixtures.Mint(pair, """{"alg":"EdDSA","typ":"JWT","kid":"self-1"}""",
                string.Create(CultureInfo.InvariantCulture,
                    $$"""{"iss":"{{Issuer}}","aud":"{{Audience}}","exp":{{Now + 3600}},"nbf":"tomorrow"}""")),
            RejectReason.ClaimsMalformed, "a non-numeric nbf is malformed, not absent");

        // NO INPUT ESCAPES Verify AS AN UNEXPECTED EXCEPTION.
        //
        // Verify documents VerifyException, so a caller writes one catch and believes it covers
        // every bad token. In the Java sibling artifact, two of these inputs made a hand-written
        // JSON reader throw an unchecked exception that sailed past the verifier's catch: in a
        // web handler that is a 500 where a 401 belonged. .NET uses System.Text.Json rather than a
        // hand-written reader, so this is checking a different thing -- that the JsonException
        // surface really is what the verifier catches, and that the depth cap holds.
        char backslash = (char)92;
        string head = "{\"alg\":\"EdDSA\",\"kid\":\"self-1\",\"x\":\"";
        string[] hostileHeaders =
        [
            head + backslash,
            head + backslash + "u12",
            head + backslash + "uZZZZ" + "\"}",
            head + backslash + "q\"}",
            string.Concat(Enumerable.Repeat("[", 6100)),
            string.Concat(Enumerable.Repeat("""{"a":""", 6100)) + "1",
        ];
        foreach (string hostile in hostileHeaders)
        {
            string token = Base64Url.Encode(Encoding.UTF8.GetBytes(hostile)) + ".e30.AA";
            _checked++;
            try
            {
                verifier.Verify(token, Now);
                Failures.Add("a hostile header VERIFIED: " + Summarise(hostile));
            }
            catch (VerifyException)
            {
                // A refusal is the point; which refusal is not, since these are all garbage.
            }
#pragma warning disable CA1031 // catching everything is the ASSERTION here, not an oversight
            catch (Exception escaped)
#pragma warning restore CA1031
            {
                Failures.Add($"a hostile header threw {escaped.GetType().Name} out of Verify(), "
                    + "which callers do not expect: " + Summarise(hostile));
            }
        }

        // THE CLAIMS DEPTH CAP IS 32, AND THAT NUMBER IS LOAD-BEARING.
        //
        // System.Text.Json already refuses runaway nesting at its own default of 64, so a
        // mutation raising the cap to 10000 SURVIVED every other check here: the 6100-deep header
        // above is refused either way. A bound nothing distinguishes is decoration, and the honest
        // options were to delete it or to measure it.
        //
        // Measured. These claims nest 40 deep: legal JSON, well inside the platform default, and
        // refused by the cap this verifier actually sets. IronAuth tokens are flat, so nothing
        // legitimate is anywhere near this.
        string deepClaims = string.Create(CultureInfo.InvariantCulture,
            $$"""{"iss":"{{Issuer}}","aud":"{{Audience}}","exp":{{Now + 3600}},"deep":""")
            + string.Concat(Enumerable.Repeat("""{"a":""", 40)) + "1" + new string('}', 40) + "}";
        Refuses(verifier, Fixtures.Mint(pair, """{"alg":"EdDSA","typ":"JWT","kid":"self-1"}""", deepClaims),
            RejectReason.ClaimsMalformed, "claims nested past the cap are refused, at 32 rather than the platform default");

        // And the control that keeps the number honest in the other direction: nesting well INSIDE
        // the cap must still verify. Without it, a cap of 1 would pass the assertion above.
        string shallowClaims = string.Create(CultureInfo.InvariantCulture,
            $$"""{"iss":"{{Issuer}}","aud":"{{Audience}}","exp":{{Now + 3600}},"deep":""")
            + string.Concat(Enumerable.Repeat("""{"a":""", 8)) + "1" + new string('}', 8) + "}";
        Accepts(verifier, Fixtures.Mint(pair, """{"alg":"EdDSA","typ":"JWT","kid":"self-1"}""", shallowClaims),
            "claims nested INSIDE the cap still verify (the control for the cap above)");

        // An empty allow-list reads as "allow nothing" and behaves as a silent outage, so it is
        // refused at construction rather than at the first request.
        _checked++;
        try
        {
            _ = new IronAuthVerifier([], keys, Issuer, Audience, 0);
            Failures.Add("an empty algorithm allow-list was accepted at construction");
        }
        catch (ArgumentException)
        {
            // as intended
        }

        // A floor, so commenting assertions out fails here instead of reporting a smaller number
        // in a green run. A floor and not an equality: adding a property should not break it.
        if (_checked < 17)
        {
            Console.Error.WriteLine($"FAIL: only {_checked} properties ran; this suite is its list");
            return 1;
        }
        if (Failures.Count > 0)
        {
            Console.Error.WriteLine("FAIL: the .NET verifier does not hold its own claims");
            Failures.ForEach(failure => Console.Error.WriteLine("  - " + failure));
            return 1;
        }
        Console.WriteLine($"dotnet self-test: {_checked} properties the corpus cannot express OK");
        return 0;
    }

    /// <summary>A short, log-safe rendering of a hostile input, so a failure names it briefly.</summary>
    private static string Summarise(string hostile) =>
        $"({hostile.Length} chars) " + (hostile.Length > 40 ? hostile[..40] + "..." : hostile);

    private static void Accepts(IronAuthVerifier verifier, string token, string why)
    {
        _checked++;
        try
        {
            verifier.Verify(token, Now);
        }
        catch (VerifyException refused)
        {
            Failures.Add($"{why} -- refused as {refused.Reason}");
        }
    }

    private static void Refuses(IronAuthVerifier verifier, string token, RejectReason expected, string why)
    {
        _checked++;
        try
        {
            verifier.Verify(token, Now);
            Failures.Add($"{why} -- but it VERIFIED");
        }
        catch (VerifyException refused)
        {
            if (refused.Reason != expected)
            {
                Failures.Add($"{why} -- expected {expected}, got {refused.Reason}");
            }
        }
    }
}
